Decimal field should be able to take Decimal values and Floats

aphillipo opened the pull request:
Hello! I've added support for the Decimal library which does arbitrary precision Decimal things.  Please review my changes.

aphillipo commented:
Fixes #90

ericmj commented on lib/xandra/protocol.ex:
```diff
@@ -306,6 +306,12 @@ defmodule Xandra.Protocol do
 
+  defp encode_value(:decimal, float) when is_float(float) do
+    encode_value(:decimal, Decimal.new(float))
+  end
```
Not sure about this casting since `:float` types exist as well. Is there casting when encoding other types?

aphillipo replied:
@ericmj if you don't think it's convenient I will remove it; just sometimes you have a float that you want to put into a decimal column.  It could allow for evil mistakes so on balance you are correct.

ericmj commented on lib/xandra/protocol.ex:
```diff
@@ -306,6 +306,12 @@ defmodule Xandra.Protocol do
+  defp encode_value(:decimal, float) when is_float(float) do
+    encode_value(:decimal, Decimal.new(float))
+  end
+  defp encode_value(:decimal, %Decimal{sign: sign, coef: coef, exp: exp}) do
```
Should probably add a guard here so we don't include exponents outside the integer range, even though it's unlikely.

aphillipo replied:
Sure thing, I will make sure of the values and get back to this.

ericmj commented:
Should we not decode to decimals also? The library is pre 1.0 so if we should do it now is the time to change it.

aphillipo commented:
Yes, I have a branch the decodes/encodes :counter vars as well as Decimals.

aphillipo commented:
@ericmj Okay so I've kept the float here, will remove it if you say to and refactor tests accordingly, the int checking I'm not too sure about the best way, I added an exp guard but the numbers are so big I thought it looked stupid so I removed it :-/  The encode actually reuses the encode_value(:int, _) function so maybe that is good enough?

Further suggestions welcome.  I'm not really sure what the alternative is here to using Decimal that some other commentators prefer?  Everyone copying the output of the decimal column into a Decimal every time they use it?

lexmag commented:
Let's not have float casting (at least for now), better to be explicit.
As far as we are at pre 1.0 stage, we can tolerate addition of decoding to `Decimal` structs.
Also we can have a compromise on adding one more dependency—make Decimal an optional dependency:
```elixir
if Code.ensure_compiled?(Decimal) do
  <decoding/encoding to %Decimal{}>
else
  <decoding/encoding to tuple>
end
```
and then we can have `{:decimal, "~> 1.0", [optional: true]}`.

aphillipo commented:
Oh I love that optional dependancy, solves issues!  Thanks for that I will implement and get back to you!

ericmj commented:
Changing between tuples and Decimals depending on if Decimal is unreliable and it should rather be a compile time option on Xandra. Otherwise the library can unexpectedly change behaviour if Decimal is added by an unrelated dependency. It can also be affected by compilation order which may not reliable.

lexmag commented:
I think we can solve this similarly to #106—provide "`:decimal_format` - (`:decimal` or `:tuple`)".
What do you all think?

lexmag commented:
Closing since there is no progress for long time.
